Create initial structure

carsongee opened the pull request:
Here is the cookiecutter

carsongee commented:
@ShawnMilo

ShawnMilo commented on {{ cookiecutter.app_name }}/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+
+# Install project packages
+RUN pip install -r requirements.txt
+RUN pip install -r test_requirements.txt
```
Why not add `-r requirements.txt` to `test_requirements.txt`?

carsongee replied:
I hadn't ever thought that you could add -r flags to files, awesome

carsongee replied:
Actually, this breaks the setup.py build, so I think I'll leave it as is.

ShawnMilo commented on {{ cookiecutter.app_name }}/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+RUN chown -R mitodl:mitodl /src
+
+# Add project
+ADD . /src
```
After this line, `ADD`ed files are owned by root, despite the directory being owned by mitodl. If this matters, moving the `chown` line down fixes that. Although there are caveats if the host environment has multiple non-root users, or the user is on a Mac.

carsongee replied:
good catch, that order was swapped around like that when I was trying to figure out the pip problem

ShawnMilo commented on {{ cookiecutter.app_name }}/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+RUN pip install -e .
+
+# Set pip cache folder, as it is breaking pip when it is on a shared volume
+ENV XDG_CACHE_HOME /tmp/.cache
```
What does this do, since it happens after the `pip install` line?

carsongee replied:
It is for during runtime, i.e. running tox installs all the packages into its own venv, so the pip caching bug mentioned applies at run time and not build time.

ShawnMilo commented on {{ cookiecutter.app_name }}/README.rst:
```diff
@@ -0,0 +1,60 @@
+up
+a near production-ready containerized development environment that
+runs migrations, with the django development server running on
+port 8070.
```
This says `8070`, but `cookiecutter.json` says `8071`.

ShawnMilo commented on {{ cookiecutter.app_name }}/README.rst:
```diff
@@ -0,0 +1,60 @@
+with
+``ptw``. This unfortunately will not work well in the docker
+environment
+because the file events it uses are fired on the host OS, and not the
```
`s/docker/Docker/`

`s/docker OS/container/`

ShawnMilo commented on {{ cookiecutter.app_name }}/apt.txt:
```diff
@@ -0,0 +1,4 @@
+libpq-dev
+python-pip
+python-dev
+python3-dev
```
The tox config contains `envlist = py27,py33,py34`, but this will never result in Python 3.3 being installed. Should `py33` be removed from that list?

carsongee replied:
py33 is still tested in travis, which runs tox, so I prefer to keep it configured.

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+from setuptools.command.test import test as testcommand
+
+with open('test_requirements.txt') as test_reqs:
+    tests_require = test_reqs.readlines()  # pylint: disable=invalid-name
```
Why not just use a different name, like `requirements`, instead of disabling the Pylint check?

carsongee replied:
that would still be a bad name, I would have to make it TEST_REQUIREMENTS or something, which I can, but since it isn't really a constant, it made more sense to me to keep it lower case.

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+    test_args = []
+
+    def initialize_options(self):
+        testcommand.initialize_options(self)
```
Isn't this function redundant?

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+        self.test_args = []
+        if self.coverage:
+            self.test_args.append('--cov')
+            self.test_args.append('{{ cookiecutter.app_name }}')
```
Might be clearer as `self.test_args.extend(['--cov', '{{ cookiecutter.app_name }}'])`.

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+        plugin_manager = _pytest.config.get_plugin_manager()
+        plugin_manager.consider_setuptools_entrypoints()
+        errno = pytest.main(self.test_args)
+        sys.exit(errno)
```
Why not just `sys.exit(pytest.main(self.test_args))` instead of the extra variable?

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+        sys.exit(errno)
+
+
+README = open('README.rst').read()
```
Missing `with` statement for cleanliness.

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+    author='MIT ODL Engineering',
+    author_email='odl-engineering@mit.edu',
+    url='http://github.com/mitodl/{{ cookiecutter.app_name }}',
+    description="{{ cookiecutter.description|replace("\"","\\\"") }}",
```
I'm not familiar with this templating language. It _looks_ like jinja2, but I don't see jinja2 in the branch, nor in docker-compose. So I can't figure out how to play with this, but is there a way to mix single quotes and double-quotes to avoid all the backslash-escaping? The line `description="{{ cookiecutter.description|replace("\"","\\\"") }}",` is not valid Python, so I assume this is fixed by the template rendering prior to being executed and perhaps that templating doesn't allow for single-quotes.

carsongee replied:
This is jinja2 via https://github.com/audreyr/cookiecutter, which runs a wizard to render everything through jinja with variables that are prompted for.  This level of escaping is required, and discovered through trial and error.  I need to turn whatever you type in the description into a string that is properly escaped for when this is an actual python file.

ShawnMilo commented on {{ cookiecutter.app_name }}/tests/test_settings.py:
```diff
@@ -0,0 +1,43 @@
+    def test_load_fallback(self):
+        """Verify our YAML load works as expected."""
+        config_settings = {'TEST_KEY': 'yessir'}
+        _, temp_config_path = tempfile.mkstemp()
```
This leaves an open filehandle. Since `os` is already imported, it would be nice to close it.

```python
handle, temp_config_path = tempfile.mkstemp()
os.close(handle)
```

ShawnMilo commented on {{ cookiecutter.app_name }}/{{ cookiecutter.app_name }}/templates/base.html:
```diff
@@ -0,0 +1,13 @@
+{% raw %}<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <title>{% block title %}{% endblock %}</title>
```
Should `<head>` contain `<meta charset="utf-8">` somewhere as the default?

ShawnMilo commented on {{ cookiecutter.app_name }}/{{ cookiecutter.app_name }}/tests/test_module.py:
```diff
@@ -0,0 +1,22 @@
+class TestModule(unittest.TestCase):
+    """
+    Test core module features, like asserting the version
+    and making sure we are exposing our classes
```
Microscopic nit: Missing period at the end of the sentence.

ShawnMilo commented on {{ cookiecutter.app_name }}/setup.py:
```diff
@@ -0,0 +1,87 @@
+        testcommand.initialize_options(self)
+
+    def finalize_options(self):
+        testcommand.finalize_options(self)
```
Just a silly question. I wasn't going to say anything about not using `super()` here, but below, in `test_views.py`, `super()` is being used. Is there a reason, or should it be consistent?

carsongee replied:
Yeah, so this is canned code from 4 projects ago.  I'll go ahead and fix it to super here, but we have a number of places we'll need to fix it

carsongee replied:
also, looking at https://pytest.org/latest/goodpractises.html there isn't really a super to call...

ShawnMilo commented on {{ cookiecutter.app_name }}/{{ cookiecutter.app_name }}/tests/test_views.py:
```diff
@@ -0,0 +1,21 @@
+    def test_index_view(self):
+        """Verify the index view is as expected"""
+        response = self.client.get(reverse('{{ cookiecutter.app_name }}_index'))
+        self.assertContains(response, 'Hi', status_code=200)
```
Nit: A two-letter word seems a bit flimsy to assert happiness, although you could argue the 200 alone is enough.

ShawnMilo commented on {{ cookiecutter.app_name }}/{{ cookiecutter.project_name }}/settings.py:
```diff
@@ -0,0 +1,146 @@
+def load_fallback():
+    """Load optional yaml config"""
+    fallback_config = {}
+    config_file_path = None
```
Minor but clearer: Either change the default value of `config_file_path` to `""`, or below check for `if config_file_path is not None`. I prefer `""`, because that way the variable is the same type at all times.

ShawnMilo commented on {{ cookiecutter.app_name }}/{{ cookiecutter.project_name }}/settings.py:
```diff
@@ -0,0 +1,146 @@
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = get_var(
+    'SECRET_KEY',
+    '36boam8miiz0c22il@3&gputb=wrqr2plah=0#0a_bknw9(2^r'
```
Since this is a template and variables are used in other places, is it easy to generate a random value here instead of hard-coded default? Many people never change their default anything.

carsongee replied:
Stock Jinja2 doesn't have a random generator filter that I'm aware of. I could at least prompt people for it to gather attention. Another similar cookiecutter project for django apps just left it as change me: https://github.com/pydanny/cookiecutter-django/blob/9f3116ba4ca4141a45e42695770d336e1a240bb3/%7B%7Bcookiecutter.repo_name%7D%7D/config/settings/local.py

ShawnMilo commented on {{ cookiecutter.app_name }}/{{ cookiecutter.project_name }}/settings.py:
```diff
@@ -0,0 +1,146 @@
+    'django.contrib.auth',
+    'django.contrib.contenttypes',
+    'django.contrib.sessions',
+    'django.contrib.messages',
```
Is `messages` used enough to be on by default? Honest question.

carsongee replied:
Enough that startproject has it, so I left it.

carsongee commented:
I believe I have addressed all your issues.  I added a second commit, as the first commit is a copy of our existing lore project, so all these changes should be applied there and having it as a second commit makes that easier to do

ShawnMilo commented:
They may be minor, and they _can_ be ignored, but these are still there:

`initialize_options` in `setup.py

ShawnMilo commented:
They may be minor, and they _can_ be ignored because they're not hurting anything, but these are still there and I didn't see comments about them.

`initialize_options` in `setup.py` is redundant.

`None` is still being used as the default value for a string in `load_fallback` in settings.py.

carsongee commented:
I thought I commented on that, I am following https://pytest.org/latest/goodpractises.html#integrating-with-distutils-python-setup-py-test from py.test which suggests it is required to be defined, even if you do nothing with it, and yeah, I prefer the concept of None and usually leave it to be implicit, but explicit is fine too.
